### PR TITLE
Run symlinked trusted Python inside candidate sandbox

### LINKED ISSUE #171: Run trusted symlinked Python interpreters inside the candidate sandbox
## Outcome

Unblock the real ChatGPT-authenticated E1 run when PMPE itself runs from a virtual environment whose `sys.executable` is a symlink to a managed Python installation.

## Real-run evidence

On the exact merged `main` commit `23189aa2b69226624006bd79c96557b5e418192f` in a local Ubuntu 26.04 Lima VM:

- ChatGPT Codex login: verified
- `/proc`, `prlimit`, and a direct `bwrap --unshare-all` smoke test: pass
- Contract compilation: `COMPILES`, approved contract and deterministic plan digests unchanged
- Provider/model call: **not reached**
- Exact sandbox error:
  `bwrap: execvp .../.venv/bin/python: No such file or directory`
- The same sandbox command succeeds with `/usr/bin/python3`.
- Copying the uv-managed standalone binary is invalid because that distribution must remain at its managed prefix.

## Root cause

`BubblewrapCandidateSandbox` binds the resolved trusted runtime roots but executes the unresolved active virtual-environment symlink. The link path is visible while its execution target is not addressable through that command path inside the sandbox.

## Acceptance criteria

- Add a failing regression proving an active `sys.executable` symlink is executed through its canonical trusted target.
- Resolve only the active trusted interpreter; do not resolve arbitrary candidate-controlled commands.
- Verify the canonical interpreter remains inside an already-bound trusted runtime root and fail closed otherwise.
- Preserve the read-only workspace, no-network namespace, cleared environment, resource limits, and existing result semantics.
- Keep the deterministic fixture and all existing candidate-isolation tests green.
- Record meaningful PR reviews, fix any findings, pass all CI, merge, and verify post-merge `main`.
- Rerun real E1 only after the merged fix is installed in the authenticated VM.

## Boundaries

No provider/API changes, no approval relaxation, no database/service/UI/deployment work, no #146 drift execution, and no API-key path.

### PR DESCRIPTION
**Pull request by Abhillashjadhav**
Closes #171.

## Live evidence and root cause

The first real E1 attempt in a ChatGPT-authenticated Lima Linux VM halted before any model call:

```
bwrap: execvp .../.venv/bin/python: No such file or directory
```

Bubblewrap succeeded with `/usr/bin/python3`, isolating the defect to the active uv virtual-environment interpreter path: the sandbox mounted the trusted runtime target but did not make every trusted interpreter alias/path component executable inside the sandbox.

## Final change

- Select runtime roots independently from the interpreter symlink target, preventing a circular trust decision.
- Strictly resolve the active trusted interpreter and require its canonical target to be a file inside the selected read-only runtime roots.
- Preserve an already mounted virtual-environment interpreter path so Python retains virtual-environment package identity.
- Map a symlinked virtual-environment directory to the same canonical trusted interpreter when that path is mounted.
- Validate the complete interpreter path-component symlink chain.
- Recreate only verified external interpreter aliases with Bubblewrap `--symlink` in otherwise empty parent directories; do not mount external directories.
- Reject reserved sandbox locations, cycles, missing links, and final escapes.
- Leave candidate-selected arbitrary command symlinks unchanged.
- Document the compatibility and security boundary.

No provider, API-key, approval-contract, evidence-schema, or platform-architecture changes are included.

## Tests-first and review evidence

The PR contains 13 incremental commits so each newly discovered edge case first appears as a failing regression test and then as its bounded fix.

Five genuine Codex review findings were accepted, fixed, replied to, and resolved:

1. circular runtime-root trust
2. loss of virtual-environment package identity
3. unmounted external interpreter aliases
4. symlinked virtual-environment directories
5. external alias components missing inside the sandbox

Final focused verification:

- `pytest tests/unit/test_candidate_sandbox.py -q` — 13 passed, 2 skipped
- `ruff format --check .`
- `ruff check .`
- `mypy --strict src`
- `bandit -q -r src -lll`
- compile-all plus generated-file diff check

The managed Work Mode host lacks the Linux `/proc`/sandbox behavior required by the complete suite, so GitHub's normal Linux CI is the full-suite gate.

## Merge and promotion gate

Merge only after:

1. exact-head Codex review completes with no unresolved finding
2. all 13 exact-head CI jobs pass
3. final exact-head owner review is recorded

After squash merge, verify `main` CI and rerun E1 in the user's ChatGPT-authenticated Lima VM. The real run uses ChatGPT subscription authentication only and no OpenAI API key.

### CODE REVIEW

**Review by Abhillashjadhav — COMMENTED** [review]
Exact-head security review for `4c7b891bda2f5ba7167960e7e68a6a2648090f61`.

Reviewed the live failure, command construction, runtime-root selection, mount mode, and regression tests. The change remains inside the existing Bubblewrap boundary: it canonicalizes only the active trusted `sys.executable`, validates the target against the same read-only roots subsequently bound into the sandbox, and leaves candidate-supplied command symlinks unresolved. The negative coverage verifies an out-of-root trusted target fails closed, while the control test verifies arbitrary command links retain their original path.

The tests-first history is auditable: the regression-only commit fails on the prior implementation and the following commit makes it pass. Focused tests plus formatting, lint, strict typing, Bandit, and compilation checks pass locally. No provider, authentication, API billing, approval, evidence, or architecture surface changed.

Owner review result: security and scope boundaries are sound; proceed to independent review and full Linux CI.

**Comment by Abhillashjadhav** [comment]
@codex review

Please review exact head `4c7b891bda2f5ba7167960e7e68a6a2648090f61`, focusing on the trust boundary: only the active `sys.executable` may be canonicalized; the resolved target must already be inside a bound read-only runtime root; candidate-selected symlinks must remain unresolved; and invalid trusted targets must fail closed. Please also verify the tests-first regression and that no provider/API/authentication or architecture surface changed.

File: src/pmpe/barebones.py [hunk 1]
```diff
@@ -165,6 +165,22 @@ def _parent_directories(path: Path) -> tuple[str, ...]:
...
+        command = list(argv)
+        if not command or command[0] != sys.executable:
+            return command
+        try:
+            canonical = Path(sys.executable).resolve(strict=True)
+        except OSError as exc:
+            raise ContractInvalidError("active Python interpreter is unavailable") from exc
+        if not canonical.is_file() or not any(
+            canonical.is_relative_to(root) for root in runtime_roots
+        ):
```
[the remark this answers is not in the document]
Comment (Abhillashjadhav) [reply src/pmpe/barebones.py:183]: Agreed—this was a circular trust check and a genuine P1 finding. Fixed in exact head `3c36fef5f2aeb82fec4f859d92a8d0959879ea8e`.

The sandbox no longer adds `Path(sys.executable).resolve().parent.parent` to its trusted roots. Allowed roots now come independently from the fixed system roots plus Python's reported `sys.base_prefix`/`sys.prefix`; only after that selection does the command target get resolved and checked for containment. A new tests-only regression (`545fabb`) proves that changing `sys.executable` to a link in an unrelated directory cannot add that directory to the trusted set. The following fix commit (`3c36fef`) makes it pass and updates the documentation.

Focused sandbox tests: 10 passed, 2 skipped. Formatting, lint, strict typing across `src`, Bandit, and compilation checks all pass.

**Comment by Abhillashjadhav** [comment]
@codex review

Please re-review exact head `3c36fef5f2aeb82fec4f859d92a8d0959879ea8e`. The P1 circular-trust finding is addressed by removing the executable-derived root; trusted roots now come independently from fixed system roots and Python base/prefix state. The tests-only commit `545fabb` proves an unrelated `sys.executable` target cannot enter the trusted set. Please verify the correction and look for any remaining security or compatibility issue.

**Review by Abhillashjadhav — COMMENTED** [review]
Post-finding security review for exact head `3c36fef5f2aeb82fec4f859d92a8d0959879ea8e`.

The P1 finding was valid and is now corrected. Trusted runtime-root selection no longer depends on the resolved `sys.executable` target. It is established first from fixed system roots and Python's base/prefix runtime state; the active command is then canonicalized only if its target is contained in that independently selected set. The new tests-only regression demonstrates that redirecting `sys.executable` into an unrelated tree does not cause that tree to be mounted or trusted. The existing positive case explicitly supplies an independently trusted managed-runtime root, while out-of-root and arbitrary-command-link cases remain fail-closed/unresolved.

Focused suite and all local static/security checks pass. Review result: the circular trust flaw is removed without broadening provider, authentication, API, or platform scope. Await independent exact-head review and full CI.

File: src/pmpe/barebones.py [hunk 2]
```diff
@@ -165,6 +163,22 @@ def _parent_directories(path: Path) -> tuple[str, ...]:
...
+            raise ContractInvalidError("active Python interpreter is outside trusted runtime roots")
+        command[0] = str(canonical)
```
[the remark this answers is not in the document]
Comment (Abhillashjadhav) [reply src/pmpe/barebones.py:206]: Agreed—this was a genuine P1 compatibility issue. Fixed in exact head `d4d918bc4b71a3433ac1877edf42fa9e03b0e2c3`.

The canonical target is still resolved strictly and checked against independently selected read-only runtime roots, but it is no longer substituted into `argv[0]`. The sandbox executes the original active virtualenv path, preserving adjacent `pyvenv.cfg` discovery and virtualenv-only packages. The positive unit case now also proves that the virtualenv root itself is part of the Bubblewrap read-only bindings. A tests-only commit (`8239520`) fails on the prior implementation by requiring the original path; the following fix makes it pass.

Focused sandbox tests: 10 passed, 2 skipped. Full formatting, lint, strict typing, Bandit, and compilation checks remain green.

**Comment by Abhillashjadhav** [comment]
@codex review

Please re-review exact head `d4d918bc4b71a3433ac1877edf42fa9e03b0e2c3`. Both prior P1 findings are addressed: trusted roots are independent of the executable target, while the canonical target is validated without replacing the original virtualenv path. The new tests-first commit `8239520` verifies virtualenv identity is retained and its root is bound read-only. Please identify any remaining security or compatibility issue.

File: src/pmpe/barebones.py [hunk 3]
```diff
@@ -165,6 +163,21 @@ def _parent_directories(path: Path) -> tuple[str, ...]:
...
+            return command
+        try:
+            canonical = Path(sys.executable).resolve(strict=True)
+        except OSError as exc:
+            raise ContractInvalidError("active Python interpreter is unavailable") from exc
+        if not canonical.is_file() or not any(
+            canonical.is_relative_to(root) for root in runtime_roots
+        ):
+            raise ContractInvalidError("active Python interpreter is outside trusted runtime roots")
+        return command
```
[the remark this answers is not in the document]
Comment (Abhillashjadhav) [reply src/pmpe/barebones.py]: Agreed. Fixed in exact head `73aac3ecc364ed94172029e8d21b32ab9de1dafb` with the conditional rule you described.

The canonical target is always resolved strictly and validated against independently selected runtime roots. The original `sys.executable` path is retained only when it is lexically covered by one of those roots (preserving an active virtualenv whose path is actually mounted). If the alias is outside every mounted root, `argv[0]` is replaced with the already-validated canonical target. Arbitrary candidate command links are still untouched.

Tests-first commit `8100a86` reproduces the unmounted-alias failure; `73aac3e` makes it pass. Focused sandbox tests are now 11 passed, 2 skipped, with formatting, lint, strict typing, Bandit, and compilation checks green.

**Comment by Abhillashjadhav** [comment]
@codex review

Please re-review exact head `73aac3ecc364ed94172029e8d21b32ab9de1dafb`. The interpreter rule is now conditional: always validate the canonical target against independently selected roots; retain the original path only when it is inside a mounted root; otherwise invoke the validated canonical target. Tests-first commit `8100a86` covers an alias outside mounted roots. Please report any remaining correctness, security, or compatibility issue.

File: src/pmpe/barebones.py [hunk 4]
```diff
@@ -165,6 +163,24 @@ def _parent_directories(path: Path) -> tuple[str, ...]:
...
+            canonical = Path(sys.executable).resolve(strict=True)
+        except OSError as exc:
+            raise ContractInvalidError("active Python interpreter is unavailable") from exc
+        if not canonical.is_file() or not any(
+            canonical.is_relative_to(root) for root in runtime_roots
+        ):
+            raise ContractInvalidError("active Python interpreter is outside trusted runtime roots")
+        active = Path(sys.executable)
+        if not any(active.is_relative_to(root) for root in runtime_roots):
+            command[0] = str(canonical)
```
[the remark this answers is not in the document]
Comment (Abhillashjadhav) [reply src/pmpe/barebones.py]: Agreed. Fixed in exact head `d472d7de4ef27f01caccfb7615024e0cc5becfc9`.

For a virtualenv whose reported `sys.prefix` is itself an alias, PMPE now derives the interpreter's path relative to that prefix and places it under the already-mounted resolved prefix. It accepts that mapped path only if it is inside a selected runtime root and resolves strictly to the same canonical interpreter already validated for trust. This preserves adjacent `pyvenv.cfg` and virtualenv-only packages without mounting the external alias or trusting an executable-derived directory. Non-virtualenv aliases still use the canonical target.

Tests-first commit `ac1789f` reproduces this symlinked-directory layout; the following fix makes it pass. Focused suite: 12 passed, 2 skipped. All formatting, lint, strict typing, Bandit, and compilation checks pass.

**Comment by Abhillashjadhav** [comment]
@codex review

Please re-review exact head `d472d7de4ef27f01caccfb7615024e0cc5becfc9`. Symlinked virtualenv directories now map the executable's prefix-relative path into the mounted resolved virtualenv and require that mapped path to resolve to the same independently validated interpreter. External aliases remain unmounted; non-virtualenv aliases use the canonical target. Tests-first commit `ac1789f` covers the layout. Please report any remaining correctness, security, or compatibility issue.

File: src/pmpe/barebones.py [hunk 5]
```diff
@@ -165,6 +163,43 @@ def _parent_directories(path: Path) -> tuple[str, ...]:
...
+                mounted_target = mounted.resolve(strict=True)
+            except OSError as exc:
+                raise ContractInvalidError("active Python virtualenv is unavailable") from exc
+            if not any(mounted.is_relative_to(root) for root in runtime_roots):
+                raise ContractInvalidError(
+                    "active Python virtualenv is outside trusted runtime roots"
+                )
+            if mounted_target != canonical:
+                raise ContractInvalidError("active Python virtualenv target is inconsistent")
+            command[0] = str(mounted)
```
[the remark this answers is not in the document]
Comment (Abhillashjadhav) [reply src/pmpe/barebones.py:204]: Agreed. Fixed in exact head `d035fa001c74b653b3d1cc3756349c186704619e`.

PMPE now walks every symlink component in the selected interpreter path, including directory aliases. Links already inside a read-only runtime root are supplied by that mount. Links outside the roots are recreated with Bubblewrap `--symlink` in otherwise empty parent directories only when the complete host chain terminates at the same independently validated canonical interpreter. Reserved sandbox paths are rejected, cycles/missing links fail closed, and no external directory contents are mounted.

Tests-first commit `e0e24f4` reproduces the dangling absolute-alias case; the following fix makes it pass. Focused suite: 13 passed, 2 skipped. Formatting, lint, strict typing, Bandit, and compilation checks pass.

**Comment by Abhillashjadhav** [comment]
@codex review

Please re-review exact head `d035fa001c74b653b3d1cc3756349c186704619e`. The sandbox now validates the complete interpreter symlink-component chain and recreates only verified external link entries in empty sandbox directories; no external content is mounted. The chain must end at the independently trusted canonical interpreter, and reserved paths/cycles/missing links fail closed. Tests-first commit `e0e24f4` covers the former dangling-alias case. Please report any remaining correctness, security, or compatibility issue.

**Review by Abhillashjadhav — COMMENTED** [review]
Final exact-head maintainer review complete. I rechecked the bounded interpreter-path design and its fail-closed security boundary after all review fixes. Five genuine Codex findings were accepted, reproduced with regression tests, fixed, replied to, and resolved. The focused sandbox suite passes 13 tests (2 real-host tests skipped locally), local format/lint/strict typing/Bandit/compile checks pass, exact-head GitHub CI is 13/13 green, the independent final Codex review on `d035fa001c` found no major issues, and there are zero unresolved review threads. Ready for squash merge; the real ChatGPT-authenticated E1 run remains a post-merge evidence gate.
